components/places/src/ffi.rs

```diff
@@ -113,6 +144,151 @@ impl PlacesConnection {
+        &self,
+        start_date: i64,
+        end_date: i64,
+        exclude_types: VisitTransitionSet,
+    ) -> Result<Vec<HistoryVisitInfo>> {
+        self.with_conn(|conn| {
+            history::get_visit_infos(
+                conn,
+                types::Timestamp(start_date.max(0) as u64),
+                types::Timestamp(end_date.max(0) as u64),
+                exclude_types,
+            )
+        })
+    }
+
+    fn get_visit_count(&self, exclude_types: VisitTransitionSet) -> Result<i64> {
+        self.with_conn(|conn| history::get_visit_count(conn, exclude_types))
+    }
+
+    fn get_visit_page(
+        &self,
+        offset: i64,
+        count: i64,
+        exclude_types: VisitTransitionSet,
+    ) -> Result<Vec<HistoryVisitInfo>> {
+        self.with_conn(|conn| history::get_visit_page(conn, offset, count, exclude_types))
+    }
+
+    fn get_visit_page_with_bound(
+        &self,
+        bound: i64,
+        offset: i64,
+        count: i64,
+        exclude_types: VisitTransitionSet,
+    ) -> Result<HistoryVisitInfosWithBound> {
+        self.with_conn(|conn| {
+            history::get_visit_page_with_bound(conn, bound, offset, count, exclude_types)
+        })
+    }
+
+    // This is identical to get_visited in history.rs but takes a list of strings instead of urls
+    // This is necessary b/c we still need to return 'false' for bad URLs which prevents us from
+    // parsing/filtering them before reaching the history layer
+    fn get_visited(&self, urls: Vec<String>) -> Result<Vec<bool>> {
+        let iter = urls.into_iter();
+        let mut result = vec![false; iter.len()];
+        let url_idxs = iter
+            .enumerate()
+            .filter_map(|(idx, s)| Url::parse(&s).ok().map(|url| (idx, url)))
+            .collect::<Vec<_>>();
+        self.with_conn(|conn| history::get_visited_into(conn, &url_idxs, &mut result))?;
+        Ok(result)
+    }
+
+    fn delete_visits_for(&self, url: Url) -> Result<()> {
+        self.with_conn(|conn| {
+            let guid = match Url::parse(url.as_str()) {
+                Ok(url) => history::url_to_guid(conn, &url)?,
+                Err(e) => {
+                    log::warn!("Invalid URL passed to places_delete_visits_for, {}", e);
+                    history::href_to_guid(conn, url.clone().as_str())?
+                }
+            };
+            if let Some(guid) = guid {
+                history::delete_visits_for(conn, &guid)?;
+            }
+            Ok(())
+        })
+    }
+
+    fn delete_visits_between(&self, start: i64, end: i64) -> Result<()> {
+        self.with_conn(|conn| {
+            history::delete_visits_between(
+                conn,
+                types::Timestamp(start.max(0) as u64),
+                types::Timestamp(end.max(0) as u64),
+            )
+        })
+    }
+
+    fn delete_visit(&self, url: Url, timestamp: i64) -> Result<()> {
+        self.with_conn(|conn| {
+            match Url::parse(url.as_str()) {
```

Comment:
This is parsing the url twice - but given the semantics, the signature of this actually needs to be `String` (the failure to parse is consumed)